### PR TITLE
structured logging

### LINKED ISSUE #84: Configure structure logging framework
## Summary
Configure our environment to use structured logging in any production environment and/or if we are emitting events to a Splunk HEC collector.

## Acceptance Criteria
- [ ] Add dep for structlog, or similar library
- [ ] Updates to logging configuration to toggle between the different formats depending on the settings.

## Background / Context
Structured logging is easier for systems like Splunk to parse, query and analyze.  Consider using https://www.structlog.org/en/stable/index.html.
https://gist.github.com/nymous/f138c7f06062b7c43c060bf03759c29e

### PR DESCRIPTION
**Pull request by ericbuckley**
## Description
Setting up the project to emit structured (JSON) logs in production so its easier to index and query in Splunk.  Also adding a request correlation id to each request, so we can group together different logging messages.

## Related Issues
closes #84 

## Additional Notes
- adding new default logging configurations for development and production (the latter includes correlation ids and outputs logs as JSON)
- adding the LOG_CONFIG env variable for selecting the logging configuration to use
- adding middleware to append a correlation id to the request (this is so we can group logs together based on the same request) and another for writing a custom access log
- removing the `recordlinker.linkage` module, as its no longer being used

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

**Review by cbrinson-rise8 — APPROVED** [review]
looks good to me!

File: Dockerfile [hunk 1]
```diff
@@ -6,6 +6,9 @@ ENV USE_OTEL=${USE_OTEL}
 # Set the USE_MSSQL env variable to true to enable SQL Server support
 ARG USE_MSSQL=true
 ENV USE_MSSQL=${USE_MSSQL}
+# Set default log config
+ARG LOG_CONFIG=/code/assets/production_log_config.json
```
Comment (alhayward) [inline Dockerfile:10]: Why do we want to set the default to be production log config, not development?

File: scripts/open_pulls.sh [hunk 2]
```diff
@@ -36,7 +36,9 @@ results=""
 while IFS= read -r pr; do
   number=$(echo "$pr" | jq -r '.number')
   # get the timestamp of the ready_for_review event
-  ready_for_review=$(gh_api issues/${number}/timeline | jq -r 'map(select(.event == "ready_for_review")) | .[0].created_at')
+  # TODO: this call assumes a PR will have less than 100 timeline events,
+  #      which is not always the case, we should handle pagination
+  ready_for_review=$(gh_api "issues/${number}/timeline?per_page=100" | jq -r 'map(select(.event == "ready_for_review")) | .[0].created_at')
```
Comment (alhayward) [inline scripts/open_pulls.sh:41]: Nice. Thanks!

File: src/recordlinker/config.py [hunk 3]
```diff
@@ -42,5 +48,53 @@ class Settings(pydantic_settings.BaseSettings):
         default="assets/initial_algorithms.json",
     )
 
+    def default_log_config(self) -> dict:
```
Comment (alhayward) [inline src/recordlinker/config.py:51]: Is there a purpose to having this default log config, vs. defining the production log config and setting as default?

**Review by alhayward — DISMISSED** [review]
Left some questions, LGTM. Is there an example of what this log will look like?

File: Dockerfile (hunk 1 above)
Comment (ericbuckley) [reply Dockerfile:10]: The Dockerfile by itself, its mostly used for customer installations, not for testing or locally development.  For example I'm guessing neither of you are running `docker build -t recordlinker .` on your laptops on a regular basis.

For development, we're using either the compose.yml (e.g. `docker compose -b run`) or just running the code natively on your machines (e.g. ./scripts/local_server.sh or uvicorn ...).

Since the Dockerfile on its own is just used for customers, it makes sense to set it appropriately.

However, I forgot that for our compose environments, using the non-json output is probably ideal!  I'll make that change in a moment.

Comment (ericbuckley) [reply Dockerfile:10]: updated.  https://github.com/CDCgov/RecordLinker/pull/93/commits/faa8aefac2f49e3dd8c65a97be250272d51e56c9

File: src/recordlinker/config.py (hunk 3 above)
Comment (ericbuckley) [reply src/recordlinker/config.py:51]: The `production_log_config.json` was created as an external asset to give customers the ability to customize that in their own installation.  If they're not happy with the defaults, they can copy that file, make adjustments and set LOG_CONFIG env variable to point to their version.

We could do this for devs to, but it just seemed like an extra file that we really didn't need to manage.  As a developer its easy to just adjust this method for a short period of time if we want to test something out.  Its easy for developers to temporarily and quickly change code, but not for customers.

**Review by cbrinson-rise8 — APPROVED** [review]
